### PR TITLE
Filestore stream-depth fixes v9

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2264

Describe changes:
This is a rework of https://github.com/OISF/suricata/pull/3423, basically a new app-layer-parser API is introduced which permits to use stream-depth for file-storing when a filestore rule is matched.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/196
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/60

### CODE REVIEW

File: src/app-layer-htp.c [hunk 1]
```diff
@@ -465,6 +465,51 @@ void AppLayerHtpNeedFileInspection(void)
...
+    if (tx_ud) {
+        if (flags & STREAM_TOCLIENT) {
+            tx_ud->tcflags |= HTP_STREAM_DEPTH_SET;
+        } else {
+            tx_ud->tsflags |= HTP_STREAM_DEPTH_SET;
+        }
+    }
+}
+
+static int AppLayerHtpCheckStreamDepth(const HtpTxUserData *tx_ud)
```
Comment (victorjulien) [inline src/app-layer-htp.c]: should the direction in which we check depend on the caller, instead of the tx_ud->t(s|c)flags?

Comment (victorjulien) [inline src/app-layer-htp.c]: could use 'bool' instead

File: src/app-layer-htp.c [hunk 2]
```diff
@@ -465,6 +465,51 @@ void AppLayerHtpNeedFileInspection(void)
...
+    if (!(flags & HTP_STREAM_DEPTH_SET) && body_limit > 0 &&
+        (content_len_so_far < body_limit) &&
+        (content_len_so_far + *len) > body_limit)
+    {
+        *len = body_limit - content_len_so_far;
+    } else if ((flags & HTP_STREAM_DEPTH_SET) && stream_depth > 0 &&
+               (content_len_so_far < stream_depth) &&
+               (content_len_so_far + *len) > stream_depth)
+    {
+        *len = stream_depth - content_len_so_far;
```
Comment (victorjulien) [inline src/app-layer-htp.c]: why return by pointer instead of returning the value?

File: src/app-layer-htp.c [hunk 3]
```diff
@@ -465,6 +465,51 @@ void AppLayerHtpNeedFileInspection(void)
...
+    }
+    return 0;
+}
+
+static void AppLayerHtpComputeChunkLength(uint64_t content_len_so_far, uint32_t body_limit,
+                                          uint32_t stream_depth, uint8_t flags, uint32_t *len)
+{
+    if (!(flags & HTP_STREAM_DEPTH_SET) && body_limit > 0 &&
+        (content_len_so_far < body_limit) &&
+        (content_len_so_far + *len) > body_limit)
```
Comment (victorjulien) [inline src/app-layer-htp.c]: I'm a bit worried with the mixing of u64 and u32 here.

File: src/app-layer-htp.c [hunk 4]
```diff
@@ -7029,6 +7081,167 @@ static int HTPParserTest25(void)
...
+
+    uint32_t len = 1000;
+
+    HtpTxUserData *tx_ud = SCMalloc(sizeof(HtpTxUserData));
+    FAIL_IF_NULL(tx_ud);
+
+    tx_ud->tsflags |= HTP_STREAM_DEPTH_SET;
+    tx_ud->request_body.content_len_so_far = 2500;
+
+    if (AppLayerHtpCheckStreamDepth(tx_ud)) {
```
Comment (victorjulien) [inline src/app-layer-htp.c]: please use no conditions like this in (new) tests. The test should always run the same path to PASS.

File: src/app-layer-htp.c (hunk 3 above)
Comment (glongo) [reply src/app-layer-htp.c]: casting `uint32_t` vars to `uint64_t` should be safer

File: src/app-layer-htp.c (hunk 1 above)
Comment (glongo) [reply src/app-layer-htp.c]: Sorry I don't get your point here

Comment (glongo) [reply src/app-layer-htp.c]: Ok understood.

**Comment by glongo** [comment]
Replaced by https://github.com/OISF/suricata/pull/3665

**Pull request closed without merging**